File: packages/shared/types/src/links.ts
/**
 * Generic Template↔Runtime Link Types
 *
 * Defines type-safe link contracts for connecting template entities
 * (CharacterInstance, ItemTemplate, etc.) to runtime entities (NPC, Item, etc.)
 * with bidirectional sync, field-level authority, and activation conditions.
 */

// Branded ID types for type safety
export type NpcId = number & { readonly __brand: 'NpcId' };
export type ItemId = number & { readonly __brand: 'ItemId' };
export type PropId = number & { readonly __brand: 'PropId' };
export type RuntimeId = NpcId | ItemId | PropId | number;

export type CharacterInstanceId = string & { readonly __brand: 'CharacterInstanceId' };
export type ItemTemplateId = string & { readonly __brand: 'ItemTemplateId' };
export type PropTemplateId = string & { readonly __brand: 'PropTemplateId' };
export type TemplateId = CharacterInstanceId | ItemTemplateId | PropTemplateId | string;

/**
 * Sync direction for template↔runtime links
 */
export type SyncDirection = 'bidirectional' | 'template_to_runtime' | 'runtime_to_template';

/**
 * Generic template↔runtime link interface
 *
 * Links a template entity (design/definition) to a runtime entity (in-game instance)
 * with configurable sync behavior and activation conditions.
 */
export interface ObjectLink {
  /** UUID primary key */
  linkId: string;

  /** Template entity type (e.g., 'characterInstance', 'itemTemplate') */
  templateKind: 'characterInstance' | 'itemTemplate' | 'propTemplate' | (string & {});

  /** Template entity ID (usually UUID) */
  templateId: TemplateId;

  /** Runtime entity type (e.g., 'npc', 'item', 'prop') */
  runtimeKind: 'npc' | 'item' | 'prop' | (string & {});

  /** Runtime entity ID (usually integer, can be UUID for some domains) */
  runtimeId: RuntimeId;

  /** Enable/disable sync for this link */
  syncEnabled: boolean;

  /** Direction of sync (bidirectional, template→runtime, or runtime→template) */
  syncDirection: SyncDirection;

  /**
   * Mapping ID pointing to registered FieldMapping config
   * Format: "templateKind->runtimeKind" (e.g., "characterInstance->npc")
   */
  mappingId?: string;

  /**
   * Priority for conflict resolution (higher priority wins)
   * Used when multiple links target the same runtime entity
   */
  priority?: number;

  /**
   * Context-based activation conditions (e.g., location, time)
   * Uses dot-notation for nested paths (e.g., {"location.zone": "downtown"})
   */
  activationConditions?: Record<string, unknown>;

  /** Extensible metadata for domain-specific use */
  meta?: Record<string, unknown>;

  /** Timestamp when link was created */
  createdAt: string;

  /** Timestamp when link was last updated */
  updatedAt: string;

  /** Timestamp of last successful sync */
  lastSyncedAt?: string;

  /** Direction of last sync */
  lastSyncDirection?: SyncDirection;
}

/**
 * Union type for all standard link kinds
 * Using ASCII-safe delimiter (->)
 */
export type LinkKind =
  | 'characterInstance->npc'
  | 'itemTemplate->item'
  | 'propTemplate->prop'
  | (string & {});

/**
 * Parsed mapping ID components
 */
export interface ParsedMappingId {
  templateKind: string;
  runtimeKind: string;
}

/**
 * Helper to extract template/runtime kinds from mapping ID
 *
 * Mapping ID format: "templateKind->runtimeKind" (e.g., "characterInstance->npc")
 *
 * @param mappingId - Mapping ID string
 * @returns Parsed template and runtime kinds
 */
export function parseMappingId(mappingId: string): ParsedMappingId {
  const parts = mappingId.split('->');
  if (parts.length !== 2) {
    throw new Error(
      `Invalid mapping ID format: "${mappingId}". Expected "templateKind->runtimeKind"`
    );
  }
  const [templateKind, runtimeKind] = parts;
  return { templateKind, runtimeKind };
}

/**
 * Type guard to check if a value is a valid ObjectLink
 */
export function isObjectLink(value: unknown): value is ObjectLink {
  if (!value || typeof value !== 'object') {
    return false;
  }

  const link = value as Partial<ObjectLink>;

  return (
    typeof link.linkId === 'string' &&
    typeof link.templateKind === 'string' &&
    typeof link.templateId === 'string' &&
    typeof link.runtimeKind === 'string' &&
    (typeof link.runtimeId === 'number' || typeof link.runtimeId === 'string') &&
    typeof link.syncEnabled === 'boolean' &&
    typeof link.syncDirection === 'string' &&
    typeof link.createdAt === 'string' &&
    typeof link.updatedAt === 'string'
  );
}

/**
 * Creates a standard mapping ID from template and runtime kinds
 *
 * @param templateKind - Template entity kind (e.g., 'characterInstance')
 * @param runtimeKind - Runtime entity kind (e.g., 'npc')
 * @returns Mapping ID string (e.g., 'characterInstance->npc')
 */
export function createMappingId(templateKind: string, runtimeKind: string): string {
  return `${templateKind}->${runtimeKind}`;
}

// =============================================================================
// API Types for Template Resolution
// =============================================================================

/**
 * Standard template kinds
 */
export type TemplateKind = 'characterInstance' | 'itemTemplate' | 'propTemplate' | (string & {});

/**
 * Standard runtime kinds
 */
export type RuntimeKind = 'npc' | 'item' | 'prop' | (string & {});

/**
 * Request to resolve a template entity to its runtime counterpart
 */
export interface ResolveTemplateRequest {
  /** Template entity kind (e.g., 'characterInstance', 'itemTemplate') */
  templateKind: TemplateKind;

  /** Template entity ID (usually UUID) */
  templateId: string;

  /**
   * Runtime context for activation-based resolution
   * Uses dot-notation paths (e.g., {'location.zone': 'downtown', 'time.period': 'night'})
   */
  context?: Record<string, unknown>;
}

/**
 * Response from template resolution
 */
export interface ResolveTemplateResponse {
  /** Whether resolution succeeded (found an active link) */
  resolved: boolean;

  /** Runtime entity kind (e.g., 'npc', 'item') - only present if resolved */
  runtimeKind?: RuntimeKind;

  /** Runtime entity ID - only present if resolved */
  runtimeId?: number;

  /** Echo of requested template kind */
  templateKind: string;

  /** Echo of requested template ID */
  templateId: string;
}

/**
 * Single item in a batch resolution request
 */
export interface ResolveBatchItem {
  templateKind: TemplateKind;
  templateId: string;
  context?: Record<string, unknown>;
}

/**
 * Request to resolve multiple template references in one call
 */
export interface ResolveBatchRequest {
  /** List of template references to resolve */
  refs: ResolveBatchItem[];

  /** Context applied to all refs (merged with per-ref context) */
  sharedContext?: Record<string, unknown>;
}

/**
 * Response from batch resolution
 */
export interface ResolveBatchResponse {
  /** Results keyed by 'templateKind:templateId' */
  results: Record<string, ResolveTemplateResponse>;

  /** Number of successfully resolved refs */
  resolvedCount: number;

  /** Total number of refs requested */
  totalCount: number;
}

/**
 * Template reference for use in interactions, scenes, etc.
 * Allows content to reference templates instead of runtime IDs.
 */
export interface TemplateRef {
  /** Template entity kind */
  templateKind: TemplateKind;

  /** Template entity ID */
  templateId: string;

  /** Optional explicit link ID to use */
  linkId?: string;
}

/**
 * Helper to create a cache key for template resolution
 */
export function createTemplateRefKey(templateKind: string, templateId: string): string {
  return `${templateKind}:${templateId}`;
}

/**
 * Parse a template ref key back to components
 */
export function parseTemplateRefKey(key: string): { templateKind: string; templateId: string } {
  const idx = key.indexOf(':');
  if (idx === -1) {
    throw new Error(`Invalid template ref key: "${key}". Expected "templateKind:templateId"`);
  }
  return {
    templateKind: key.slice(0, idx),
    templateId: key.slice(idx + 1),
  };
}

File: pixsim7/backend/main/domain/links.py
"""Generic Template↔Runtime Link Domain Models

This module provides generic link infrastructure for connecting any template entity
(CharacterInstance, ItemTemplate, PropTemplate) to any runtime entity (NPC, Item, Prop).

The link pattern supports:
- Bidirectional sync with field-level authority
- Priority-based conflict resolution
- Context-based activation (e.g., location, time)
- Mapping configuration via registry
- Extensible metadata

Architecture:
    Template Entity (characterInstance, itemTemplate, etc.)
      └─ ObjectLink
           └─ Runtime Entity (npc, item, etc.)

Example links:
- characterInstance->npc: CharacterInstance ↔ GameNPC
- itemTemplate->item: ItemTemplate ↔ GameItem
- propTemplate->prop: PropTemplate ↔ PropInstance
"""
from datetime import datetime
from typing import Optional, Dict, Any
from uuid import UUID, uuid4
from sqlmodel import SQLModel, Field, Column
from sqlalchemy.dialects.postgresql import JSONB
from sqlalchemy import Index

from pixsim7.backend.main.shared.datetime_utils import utcnow


class ObjectLink(SQLModel, table=True):
    """Generic template↔runtime link table

    Links a template entity (design/definition) to a runtime entity (in-game instance)
    with configurable sync behavior, priority, and activation conditions.

    Mapping ID format: "templateKind->runtimeKind" (e.g., "characterInstance->npc")

    Example:
        from pixsim7.backend.main.services.links.link_types import link_type_id

        ObjectLink(
            template_kind='characterInstance',
            template_id='abc-123-uuid',
            runtime_kind='npc',
            runtime_id=456,
            mapping_id=link_type_id('characterInstance', 'npc'),
            priority=10,
            activation_conditions={'location.zone': 'downtown'}
        )
    """
    __tablename__ = "object_links"

    # Primary key
    link_id: UUID = Field(default_factory=uuid4, primary_key=True)

    # Template reference
    template_kind: str = Field(
        max_length=50,
        description="Template entity kind (e.g., 'characterInstance', 'itemTemplate')"
    )
    template_id: str = Field(
        max_length=255,
        description="Template entity ID (usually UUID)"
    )

    # Runtime reference
    runtime_kind: str = Field(
        max_length=50,
        description="Runtime entity kind (e.g., 'npc', 'item', 'prop')"
    )
    runtime_id: int = Field(
        description="Runtime entity ID (usually integer)"
    )

    # Sync configuration
    sync_enabled: bool = Field(
        default=True,
        description="Enable/disable sync for this link"
    )
    sync_direction: str = Field(
        default='bidirectional',
        max_length=50,
        description="Sync direction: 'bidirectional', 'template_to_runtime', 'runtime_to_template'"
    )

    # Mapping reference
    mapping_id: Optional[str] = Field(
        None,
        max_length=100,
        description="Mapping ID pointing to FieldMapping config (e.g., 'characterInstance->npc')"
    )

    # Per-link sync field mappings (simple path-to-path map for sync operations)
    # Format: {"source.path": "target.path"} e.g., {"visual_traits.scars": "personality.appearance.scars"}
    # This is for sync services, NOT for overriding FieldMapping registry configs
    sync_field_mappings: Optional[Dict[str, str]] = Field(
        None,
        sa_column=Column(JSONB),
        description="Per-link field mappings for sync operations (source_path -> target_path)"
    )

    # Link behavior
    priority: int = Field(
        default=0,
        description="Priority for conflict resolution (higher priority wins)"
    )
    activation_conditions: Optional[Dict[str, Any]] = Field(
        None,
        sa_column=Column(JSONB),
        description="Context-based activation conditions (e.g., {'location.zone': 'downtown'})"
    )

    # Metadata
    meta: Optional[Dict[str, Any]] = Field(
        None,
        sa_column=Column(JSONB),
        description="Extensible metadata for domain-specific use"
    )
    created_at: datetime = Field(default_factory=utcnow)
    updated_at: datetime = Field(default_factory=utcnow)
    last_synced_at: Optional[datetime] = Field(None)
    last_sync_direction: Optional[str] = Field(None, max_length=50)

    # Table args for indexes
    __table_args__ = (
        Index('ix_object_links_template', 'template_kind', 'template_id'),
        Index('ix_object_links_runtime', 'runtime_kind', 'runtime_id'),
        Index('ix_object_links_mapping_id', 'mapping_id'),
        Index('ix_object_links_priority', 'runtime_kind', 'runtime_id', 'priority'),
    )

    class Config:
        arbitrary_types_allowed = True

    def __repr__(self) -> str:
        return (
            f"ObjectLink("
            f"{self.template_kind}:{self.template_id} -> "
            f"{self.runtime_kind}:{self.runtime_id}, "
            f"mapping={self.mapping_id}, "
            f"priority={self.priority})"
        )
